src/Servers/Kestrel/Core/src/Internal/Http2/Http2Connection.cs

```diff
@@ -1087,7 +1098,6 @@ private void MakeSpaceInDrainQueue()
                 }
 
                 _streams.Remove(stream.StreamId);
-                ReturnStream(stream);
```

Comment:
This call to ReturnStream can also stay. While It's true that CompleteStream() also eventually causes a stream to be returned to a pool, that's a different stream. This is returning a stream that used to be in the _completedStreams queue.

Reply:
Hmmm, are you sure? You'd want to have the same check as up above: Has the stream received END_STREAM or RST_STREAM.

I'd rather be safe and say that if we're removing streams on an unexpected path we shouldn't try to reuse them. Better to be safe than sorry.

Reply:
Re-added